File: python/test/t_MeasureFactory_std.py
import openturns as ot
import openturns.testing as ott
import otrobopt

full_fun = ot.SymbolicFunction(["x", "theta"], ["theta*x"])
param_fun = ot.ParametricFunction(full_fun, [1], [1.0])

# Normal distribution associated to parameter theta
thetaDist = ot.Normal(2.0, 0.1)

# Definition of "analytical" mean measure associated to the parametric function
mean_measure = otrobopt.MeanMeasure(param_fun, thetaDist)
variance_measure = otrobopt.VarianceMeasure(param_fun, thetaDist)
# Definition of "discretized" mean measure
N = 10
experiment = ot.LHSExperiment(thetaDist, N)
factory = otrobopt.MeasureFactory(experiment)
discretized_mean_measure_evaluation = factory.build(mean_measure)

coll = [mean_measure, variance_measure]
discretizedColl = factory.buildCollection(coll)
ott.assert_almost_equal(len(coll), 2)

x = [1.0]
ott.assert_almost_equal(otrobopt.AggregatedMeasure(coll)(x), [2.0, 0.01])

File: lib/test/t_MeasureFactory_std.cxx
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"
#include "otrobopt/OTRobOpt.hxx"

using namespace OT;
using namespace OT::Test;
using namespace OTROBOPT;

int main()
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    Normal thetaDist(2.0, 0.1);
    SymbolicFunction f_base(Description({"x", "theta"}), Description({"x*theta"}));
    ParametricFunction f(f_base, Indices(1, 1), Point(1, 1.0));

    Point x(1, 1.0);

    // Test MeasureFactory build single measure
    MeasureEvaluation mean(MeanMeasure(f, thetaDist));
    LHSExperiment lhs(200);
    MeasureFactory factory(lhs);
    MeasureEvaluation discretized(factory.build(mean));
    fullprint << "discretized mean=" << discretized(x) << std::endl;

    // Test buildCollection
    Collection<MeasureEvaluation> measures;
    measures.add(MeanMeasure(f, thetaDist));
    measures.add(VarianceMeasure(f, thetaDist));

    MeasureFactory::MeasureEvaluationCollection coll(factory.buildCollection(measures));
    fullprint << "collection size=" << coll.getSize() << std::endl;
    fullprint << "coll[0]=" << coll[0](x) << std::endl;
    fullprint << "coll[1]=" << coll[1](x) << std::endl;
    assert_almost_equal(coll.getSize(), 2);

    // All items in a collection share the same distribution
    Distribution dist0(coll[0].getDistribution());
    Distribution dist1(coll[1].getDistribution());
    assert_equal(dist0.getImplementation().get(), dist1.getImplementation().get());
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }
  return ExitCode::Success;
}
